### ISSUE TITLE
Statefile is deleted when I tried to rename a stack. I don't know how to reproduce.

### ISSUE THREAD

**Issue by mike-chen-samsung** [pulumi/pulumi#8896]
## Hello!

- Vote on this issue by adding a 👍 reaction
- To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already)

## Issue details

Statefile is deleted when I tried to rename a stack. I had to recover from the `.bak` statefile.

```
$ pulumi stack rename dev/global-variables
================================================================================
The Pulumi CLI encountered a fatal error. This is a bug!
We would appreciate a report: https://github.com/pulumi/pulumi/issues/
Please provide all of the below text in your report.
================================================================================
Pulumi Version:   v3.23.2
Go Version:       go1.17.6
Go Compiler:      gc
Architecture:     amd64
Operating System: darwin
Panic:            runtime error: slice bounds out of range [-1:]

goroutine 1 [running]:
runtime/debug.Stack()
        /usr/local/Cellar/go/1.17.6/libexec/src/runtime/debug/stack.go:24 +0x65
main.panicHandler()
        /private/tmp/pulumi-20220128-33057-ihclag/pkg/cmd/pulumi/main.go:29 +0x3d
panic({0x55ec060, 0xc000505008})
        /usr/local/Cellar/go/1.17.6/libexec/src/runtime/panic.go:1038 +0x215
github.com/pulumi/pulumi/pkg/v3/backend/filestate.(*localBackend).renameHistory(0xc0004d1980, {0xc000116420, 0xc000049230}, {0x7ffeefbff7f4, 0x14})
        /private/tmp/pulumi-20220128-33057-ihclag/pkg/backend/filestate/state.go:406 +0x465
github.com/pulumi/pulumi/pkg/v3/backend/filestate.(*localBackend).RenameStack(0xc0004d1980, {0x5a92c28, 0xc0001b4008}, {0x5acae28, 0xc000b78f90}, {0x7ffeefbff7f4, 0x14})
        /private/tmp/pulumi-20220128-33057-ihclag/pkg/backend/filestate/backend.go:429 +0x2e5
github.com/pulumi/pulumi/pkg/v3/backend.RenameStack({0x5a92c28, 0xc0001b4008}, {0x5acae28, 0xc000b78f90}, {0x7ffeefbff7f4, 0x14})
        /private/tmp/pulumi-20220128-33057-ihclag/pkg/backend/stack.go:72 +0x78
github.com/pulumi/pulumi/pkg/v3/backend/filestate.(*localStack).Rename(0xc000116420, {0x5a92c28, 0xc0001b4008}, {0x7ffeefbff7f4, 0x0})
        /private/tmp/pulumi-20220128-33057-ihclag/pkg/backend/filestate/stack.go:64 +0x3e
main.newStackRenameCmd.func1(0x0, {0xc000133a00, 0x1, 0xc00017bd78})
        /private/tmp/pulumi-20220128-33057-ihclag/pkg/cmd/pulumi/stack_rename.go:65 +0x142
github.com/pulumi/pulumi/sdk/v3/go/common/util/cmdutil.RunFunc.func1(0xc, {0xc000133a00, 0xc00017bd90, 0x41e9073})
        /private/tmp/pulumi-20220128-33057-ihclag/sdk/go/common/util/cmdutil/exit.go:96 +0x26
github.com/pulumi/pulumi/sdk/v3/go/common/util/cmdutil.RunResultFunc.func1(0xc000a97b80, {0xc000133a00, 0x1, 0x1})
        /private/tmp/pulumi-20220128-33057-ihclag/sdk/go/common/util/cmdutil/exit.go:112 +0x4f
github.com/spf13/cobra.(*Command).execute(0xc000a97b80, {0xc0001339e0, 0x1, 0x1})
        /Users/brew/Library/Caches/Homebrew/go_mod_cache/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:846 +0x5f8
github.com/spf13/cobra.(*Command).ExecuteC(0xc000588000)
        /Users/brew/Library/Caches/Homebrew/go_mod_cache/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:950 +0x3ad
github.com/spf13/cobra.(*Command).Execute(...)
        /Users/brew/Library/Caches/Homebrew/go_mod_cache/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:887
main.main()
        /private/tmp/pulumi-20220128-33057-ihclag/pkg/cmd/pulumi/main.go:48 +0x45
```

### Steps to reproduce

I'm not sure. But here's my bash history in case it helps. I am using `--local` state store

```
  502  pulumi stack init --secrets-provider="awskms://alias/pulumi-platform-infra"  <--- I forget what I named the stack initially here. Something like `global-variables-dev` maybe?
  503  pulumi config set --secret DATADOG_API_KEY <redacted>
  504  pulumi stack rename dev
  505  pwd
  506  pulumi preview
  507  pulumi preview
  508  pulumi up
  509  pulumi up
  510  pulumi up
  511  pulumi stack
  512  pulumi stack
  513  pulumi stack rename bixby/global-variables/dev
  514  pulumi stack
  515  pulumi preview
  516  pulumi stack rename dev/global-variables <--- ERROR here
```

**Comment by Frassle** [comment]
The local backend doesn't currently support the concept of org or project. I wonder if this is a bug where it isn't validating correctly that "dev/global-variables" isn't valid for the local backend (because of the "dev/") and just passing that on to filesystem calls and getting confused because folders do/don't exist.

**Comment by mike-chen-samsung** [comment]
I've tried all sorts of combinations of switching between the stack names below without error. But I only did this for a new empty stack, not for a stack with things in it. I mainly submitted this ticket because the error message asked me to.

```
hi/hi/hi
hi/hi
hi
hi-hi-hi
hi-hi
```

Does only Pulumi cloud support the concept of orgs and projects, or do Cloud backends like S3 support it too? It might be good to be clearer about that in the docs. All I can find regarding the stack name format is

> Fully qualified stack names also include the project name, in the form orgName/projectName/stackName, and this fully-qualified format is required in some contexts

Which doesn't mention anything about backend compatibility.
